plugins/advisors/black-duck/src/main/kotlin/BlackDuck.kt

```diff
@@ -0,0 +1,205 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.plugins.advisors.blackduck
+
+import com.blackduck.integration.blackduck.api.generated.view.OriginView
+import com.blackduck.integration.blackduck.api.generated.view.VulnerabilityView
+
+import java.time.Instant
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.withContext
+
+import org.apache.logging.log4j.kotlin.logger
+
+import org.ossreviewtoolkit.advisor.AdviceProvider
+import org.ossreviewtoolkit.advisor.AdviceProviderFactory
+import org.ossreviewtoolkit.model.AdvisorCapability
+import org.ossreviewtoolkit.model.AdvisorDetails
+import org.ossreviewtoolkit.model.AdvisorResult
+import org.ossreviewtoolkit.model.AdvisorSummary
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.Issue
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.Severity
+import org.ossreviewtoolkit.model.createAndLogIssue
+import org.ossreviewtoolkit.model.vulnerabilities.Cvss2Rating
+import org.ossreviewtoolkit.model.vulnerabilities.Vulnerability
+import org.ossreviewtoolkit.model.vulnerabilities.VulnerabilityReference
+import org.ossreviewtoolkit.plugins.api.OrtPlugin
+import org.ossreviewtoolkit.plugins.api.PluginDescriptor
+import org.ossreviewtoolkit.utils.common.enumSetOf
+
+@OrtPlugin(
+    id = "BlackDuck",
+    displayName = "BlackDuck",
+    description = "An advisor that retrieves vulnerability information from a BlackDuck instance.",
+    factory = AdviceProviderFactory::class
+)
+class BlackDuck(
+    override val descriptor: PluginDescriptor,
+    private val blackDuckApi: ComponentServiceClient
+) : AdviceProvider {
+    override val details = AdvisorDetails(descriptor.id, enumSetOf(AdvisorCapability.VULNERABILITIES))
+
+    constructor(descriptor: PluginDescriptor, config: BlackDuckConfiguration) : this(
+        descriptor, ExtendedComponentService.create(config.serverUrl, config.apiToken)
+    )
+
+    override suspend fun retrievePackageFindings(packages: Set<Package>): Map<Package, AdvisorResult> {
+        val startTime = Instant.now()
+        val issuesForId = packages.associateBy({ it.id }, { mutableListOf<Issue>() })
+
+        logger.info { "Obtaining origins for ${packages.size} packages..." }
+
+        val originsForId = withContext(Dispatchers.IO) {
```

Comment:
I guess this should use `Dispatchers.IO.limitedParallelism(20)`, see 23c9bb0d018f92a4444167293e677ae8c5f6f561. Right @mnonnenmacher?

Reply:
I believe limiting parallelization is an unnecessary limitation in case of the advisor. because there is not so much other stuff going on in parallel as in the analyzer. IMO it's fine as is. Agree?

Reply:
I'm not quite convinced yet, as multiple other advisors could run in parallel. On the other hand, the other advisor implementations do not seem to use `Dispatchers.IO`.

Reply:
Ok, I added it.